File: PERL/Cap08/c08ex10.pl
# c08ex10.pl

#!/usr/bin/perl

use strict;
use warnings;

sub fibonacciBase {
  my ($N, $X, $Y) = @_;
  if ($N == 0) {
    return $X;
  }
  if ($N == 1) {
    return $Y;
  }
  return fibonacciBase($N - 1, $Y, $X + $Y);
}

sub FIBONACCI {
  my ($N) = @_;
  return fibonacciBase($N, 0, 1);
}

for (my $I = 0; $I <= 14; $I++) {
  printf "%2d - %10d\n", $I, FIBONACCI($I);
}

print "\n";
print "Tecle <Enter> para encerrar... ";
<STDIN>;

File: GO/Cap08/c08ex10.go
// c08ex10.go

package main

import "fmt"

func FibonacciBase(N int, X int, Y int) int {
  if N == 0 {
    return X
  }
  if N == 1 {
    return Y
  }
  if N == 2 {
    return X + Y
  }
  if N >= 2 {
    return FibonacciBase(N-1, Y, X+Y)
  }
  return 0
}

func Fibonacci(N int) int {
  return FibonacciBase(N, 0, 1)
}

func main() {
  for I := 0; I <= 14; I++ {
    fmt.Printf("%2d - %d\n", I, Fibonacci(I))
  }
  fmt.Println()
  fmt.Print("Tecle <Enter> para encerrar... ")
  fmt.Scanln()
}
